### PR TITLE
star 2023-09-28

### PR DESCRIPTION
**Pull request by alebcay**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

Would appreciate some help on adding a livecheck. The version `1.7.0` is derived from file `star/version.c` while the date portion comes from the tag.

### CODE REVIEW

**Review by p-linnane — APPROVED** [review]

File: Formula/s/star.rb [hunk 1]
```diff
@@ -1,8 +1,9 @@
 class Star < Formula
   desc "Standard tap archiver"
-  homepage "https://cdrtools.sourceforge.net/private/star.html"
-  url "https://downloads.sourceforge.net/project/s-tar/star-1.5.3.tar.bz2"
-  sha256 "070342833ea83104169bf956aa880bcd088e7af7f5b1f8e3d29853b49b1a4f5b"
+  homepage "https://codeberg.org/schilytools/schilytools"
+  url "https://codeberg.org/schilytools/schilytools/archive/2023-09-28.tar.gz"
+  version "1.7.0-2023-09-28"
+  sha256 "564ea2365876a53eba02f184c565016399aee188c26d862589906cf3f92198e6"
   license "CDDL-1.0"
```
Comment (samford) [inline Formula/s/star.rb:8]: If we want to check the Git tags:

```suggestion

  livecheck do
    url :stable
    regex(/^(\d+(?:[.-]\d+)+)$/i)
    strategy :git do |tags, regex|
      # Find the highest version tag
      tag = tags.map { |tag| tag[regex, 1] }
                .flatten
                .uniq
                .map { |v| Version.new(v) }
                .sort
                .last
      next if tag.blank?

      # Fetch the `star/version.c` file for the highest version tag
      version_file = Homebrew::Livecheck::Strategy.page_content(
        "https://codeberg.org/schilytools/schilytools/raw/tag/#{tag}/star/version.c",
      )
      next if version_file[:content].blank?

      # Identify the `star` version from the `version.c` file
      version = version_file[:content][/strvers\[\]\s*=\s*"([^"]+?)"/i, 1]
      next if version.blank?

      "#{version}-#{tag}"
    end
  end

```

If we want to check release tags:

```suggestion

  livecheck do
    url "https://codeberg.org/schilytools/schilytools/releases"
    regex(%r{href=.*?/src/tag/(\d+(?:[.-]\d+)+)/?["' >]}i)
    strategy :page_match do |page, regex|
      # Find the newest release tag
      tag = page.scan(regex).map { |match| match[0] }&.first
      next if tag.blank?

      # Fetch the `star/version.c` file for the newest release tag
      version_file = Homebrew::Livecheck::Strategy.page_content(
        "https://codeberg.org/schilytools/schilytools/raw/tag/#{tag}/star/version.c",
      )
      next if version_file[:content].blank?

      # Identify the `star` version from the `version.c` file
      version = version_file[:content][/strvers\[\]\s*=\s*"([^"]+?)"/i, 1]
      next if version.blank?

      "#{version}-#{tag}"
    end
  end

```

**Review by samford — COMMENTED** [review]
Hmm, if the `star/version.c` file is the only source of the `1.7.0` part of the formula version, then we'll have to make multiple requests to assemble the information. We can identify the highest Git tag or the newest release tag, fetch the `version.c` file using that tag, and match the version from the file. I've added two suggestions below for these approaches.

Looking at recent versions on [Repology](https://repology.org/project/star-tar/versions), MacPorts is using a version like `1.7.0-2023-04-19` and FreeBSD Ports is simply using `2023.09.28`. The aforementioned approaches work if we really want the `1.7.0-2023-09-28` version format. Otherwise, if the date is fine, we could simply check the tags (or releases).

File: Formula/s/star.rb (hunk 1 above)
Comment (alebcay) [reply Formula/s/star.rb:8]: Thanks for helping flush both options out, possibly also pertinent is upstream discussion I raised at https://codeberg.org/schilytools/schilytools/issues/61.

**Comment by alebcay** [comment]
Thanks for helping flush both options out, possibly also pertinent (regarding version scheme) is upstream discussion I raised at [codeberg.org/schilytools/schilytools/issues/61](https://codeberg.org/schilytools/schilytools/issues/61).

**Comment by samford** [comment]
Thinking about this some more, one challenging aspect about the upstream release situation is that `star` may not have changed in a particular release. Also, as you mentioned in the issue, the date in the `version.c` file isn't updated for all changes.

The current date in `star/version.c` is 2023/01/11 (the subsequent release is 2023-01-12) but there was a change to `star` in the 2023-04-19 release:

> star: remove some C99-only code accidentally introduced in the previous release.

This change may not have been seen as significant enough to warrant changing the date in `version.c` but it's a change nonetheless. The 2023-09-28 release doesn't seem to contain any additional changes to `star` but the FreeBSD port still uses it.

Ideally, we would only update the formula to a new release if there are changes to `star` but identifying that in `livecheck` requires some additional logic. It would be the same general idea as the aformentioned approaches but we would identify the tag with a date that's greater than or equal to the date in the `version.c` file from the latest tag.

If we naively use the newest release then we may periodically do unnecessary version bumps but if we use the date in `version.c` to find the closest release we may miss minor changes. So long as these minor changes don't affect the behavior of the application, I think there's something to be said about being conservative about version bumps (unless/until it causes problems).

I'll add some suggestions for these approaches as well.

File: Formula/s/star.rb (hunk 1 above)
Comment (samford) [reply Formula/s/star.rb:8]: If we want to check the Git tags and use the tag that's closest to the date in `version.c`:

```suggestion

  livecheck do
    url :stable
    regex(/^(\d+(?:[.-]\d+)+)$/i)
    strategy :git do |tags, regex|
      # Find the highest version tag
      vtags = tags.map { |tag| tag[regex, 1] }
                  .compact
                  .map { |tag| Version.new(tag) }
      next if vtags.blank?

      # Fetch the `star/version.c` file for the highest version tag
      latest_tag = vtags.max.to_s
      version_file = Homebrew::Livecheck::Strategy.page_content(
        "https://codeberg.org/schilytools/schilytools/raw/tag/#{latest_tag}/star/version.c",
      )
      next if version_file[:content].blank?

      # Identify the `star` version and date from the `version.c` file
      version = version_file[:content][/strvers\[\]\s*=\s*"([^"]+?)"/i, 1]
      date = version_file[:content][/dvers\[\]\s*=\s*"([^"]+?)"/i, 1]
      next if version.blank? || date.blank?

      # Identify the tag that's closest to the version date
      date_version = Version.new(date)
      tag = vtags.select { |v| v >= date_version }.min

      "#{version}-#{tag}"
    end
  end

```

If we want to check release tags and use the tag that's closest to the date in `version.c`:

```suggestion

  livecheck do
    url "https://codeberg.org/schilytools/schilytools/releases"
    regex(%r{href=.*?/src/tag/(\d+(?:[.-]\d+)+)/?["' >]}i)
    strategy :page_match do |page, regex|
      # Find the highest release version tag
      vtags = page.scan(regex).map { |match| match[0] }
                  .map { |tag| Version.new(tag) }
      next if vtags.blank?

      # Fetch the `star/version.c` file for the highest release tag
      latest_tag = vtags.max.to_s
      version_file = Homebrew::Livecheck::Strategy.page_content(
        "https://codeberg.org/schilytools/schilytools/raw/tag/#{latest_tag}/star/version.c",
      )
      next if version_file[:content].blank?

      # Identify the `star` version and date from the `version.c` file
      version = version_file[:content][/strvers\[\]\s*=\s*"([^"]+?)"/i, 1]
      date = version_file[:content][/dvers\[\]\s*=\s*"([^"]+?)"/i, 1]
      next if version.blank? || date.blank?

      # Identify the tag that's closest to the version date
      date_version = Version.new(date)
      tag = vtags.select { |v| v >= date_version }.min

      "#{version}-#{tag}"
    end
  end

```

**Comment by alebcay** [comment]
Sure, that's also an option. I just ask that whatever we decide here, we also be consistent in applying to `smake`, where we've already gone in the direction that this PR currently is going.

Note that `star` doesn't just depend on only contents in the `star` directory - it also uses code from other folders in the repo, including `librmt`, `libdeflt`, `libfind`, `libschily` (these are built into static libs and then added into `star`). The date in `version.c` may not capture these changes. I'm not sure these libraries are versioned like `smake`/`star` either - date might be the only way to track these.

We may also want to consider what to do about `cdrtools`, which is now also part of `schilytools` but is comprised of multiple utilities in multipel folders in the source tree, unlike `star` and `smake`. Not sure what the versioning situation looks like there.

Another option to think about is finding a way to add support upstream for dynamic libraries. If we decide to move `librmt`, `libdeflt`, `libfind`, `libschily`, etc. into their own formulae, then maybe just a date-only versioning scheme makes more sense.

**Comment by samford** [comment]
> I just ask that whatever we decide here, we also be consistent in applying to `smake`, where we've already gone in the direction that this PR currently is going.

Depending on what we do here, `smake` may need a `livecheck` block as well, so I'm happy to create a follow-up PR after we're done here 👍

> Note that `star` doesn't just depend on only contents in the `star` directory - it also uses code from other folders in the repo, including `librmt`, `libdeflt`, `libfind`, `libschily` (these are built into static libs and then added into `star`). The date in `version.c` may not capture these changes. I'm not sure these libraries are versioned like `smake`/`star` either - date might be the only way to track these.

Ah, thanks for the clarification. In that case, it's probably too much effort to identify when `star` is affected by the changes in a given release and we'll just have to accept that some version bumps may not be meaningful (this sometimes happens in other formulae, so it's nothing new).

> We may also want to consider what to do about `cdrtools`, which is now also part of `schilytools` but is comprised of multiple utilities in multipel folders in the source tree, unlike `star` and `smake`. Not sure what the versioning situation looks like there.

Repology looks similar for `cdrtools`, where MacPorts is using a 3.02-2023-04-19 version format and FreeBSD is using 2023.09.28. Looking through the source, the easiest way to get the 3.02 part of the version would be to check [`cdrecord/version.h`](https://codeberg.org/schilytools/schilytools/src/tag/2023-09-28/cdrecord/version.h).

> Another option to think about is finding a way to add support upstream for dynamic libraries. If we decide to move `librmt`, `libdeflt`, `libfind`, `libschily`, etc. into their own formulae, then maybe just a date-only versioning scheme makes more sense.

I suppose the question is whether we use a multipart version like 1.7.0-2023-09-28 in the interim time or simply opt for a date version now (2023-09-28 or 2023.09.28). If we're going to be updating related formulae for every schilytools release anyway, then using a date-based version would allow us to add the formulae to `synced_versions_formulae.json`. We also wouldn't have to add `livecheck` blocks to the formulae (for the time being) unless/until we need to check releases (instead of tags) or massage the version format (even then, it would be much simpler than the suggestions above).

**Comment by alebcay** [comment]
I think you raise some good points about the date-only versioning; simpler livecheck and synced formulae seem like good enough reasons. Since upstream is doing date-based releases for now, it seems more consistent, compared to some of the ideas we bounced around about trying to identify the last change that touched a tool, etc.

File: Formula/s/star.rb [hunk 2]
```diff
@@ -1,8 +1,9 @@
 class Star < Formula
   desc "Standard tap archiver"
-  homepage "https://cdrtools.sourceforge.net/private/star.html"
-  url "https://downloads.sourceforge.net/project/s-tar/star-1.5.3.tar.bz2"
-  sha256 "070342833ea83104169bf956aa880bcd088e7af7f5b1f8e3d29853b49b1a4f5b"
+  homepage "https://codeberg.org/schilytools/schilytools"
+  url "https://codeberg.org/schilytools/schilytools/archive/2023-09-28.tar.gz"
+  version "1.7.0-2023-09-28"
```
Comment (samford) [inline Formula/s/star.rb:5]:
```suggestion
  version "2023-09-28"
```

**Review by samford — COMMENTED** [review]
I opened a brew PR (https://github.com/Homebrew/brew/pull/16081) to tweak the date version regex, so it will also handle file names without a prefix (e.g., `2023-09-28.tar.gz` instead of only `something-2023-09-28.tar.gz`). That will work for `smake` and `star` but I think we'll have to wait until it's available in a new `brew` release before we can take advantage of it.

I have some changes stashed for related formulae, so we can move forward with a manually-specified date version for now and I'll simply update this with the others.

**Review by SMillerDev — APPROVED** [review]

**Pull request merged**